### DISCUSSION TITLE
Create a bridge between ROS and PyBullet

### DISCUSSION THREAD

**Discussion by erwincoumans** [bulletphysics/bullet3#3597] [General]
See also https://pybullet.org/Bullet/phpBB3/viewtopic.php?f=24&t=12623

**Comment by erwincoumans** [comment]
As alternative, there is Gibson, which uses PyBullet and its own rendering. It seems to be ROS bindings: https://github.com/StanfordVL/GibsonEnv

**Comment by QingkaiLu** [comment]
It would be convenient for people who use ROS to have a bridge between ROS and PyBullet.

**Comment by mbusy** [comment]
I agree, it would be very convenient. In our project [qibullet](https://github.com/ProtolabSBRE/qibullet) we ended up creating a wrapper class, creating a bridge between ROS and our simulated model

**Comment by oscar-lima** [comment]
I have started a pybullet ROS wrapper [here](https://github.com/oscar-lima/pybullet_ros), still in early stage but help and coments are welcome ; )

**Comment by erwincoumans** [comment]
Thanks @oscar-lima and self-contained pybullet ROS wrapper would be nice. What features do you plan to wrap? I imagine there could be control (setJointMotorControl2), sensors (joint angle encoder), camera vision (getCameraImage), LIDAR (using rayTestBatch batch ray cast, see [example code](https://github.com/erwincoumans/pybullet_robots/blob/master/f10_racecar.py#L131)), touch/contact (using getContactPoints).

**Comment by erwincoumans** [comment]
Also related to ROS+PyBullet is this project:
https://github.com/ProtolabSBRE/qibullet

**Comment by oscar-lima** [comment]
> What features do you plan to wrap?

Control and sensors was on my list, but I welcome your suggestions. Thanks!

**Comment by oscar-lima** [comment]
> I have started a pybullet ROS wrapper [here](https://github.com/oscar-lima/pybullet_ros), still in early stage but help and comments are welcome ; )

@erwincoumans, basic implementation is done, more urgently what I am missing is the parallelization of the [plugins](https://github.com/oscar-lima/pybullet_ros/tree/melodic/ros/src/pybullet_ros/plugins) in threads. Can you please take a look and provide some feedback? (main loop [here](https://github.com/oscar-lima/pybullet_ros/blob/melodic/ros/src/pybullet_ros/pybullet_ros.py#L225))

**Comment by erwincoumans** [comment]
Thanks for the update, that looks great from a first glance!

What do mean by parallelization? Is this a C plugin for PyBullet or a Python plugin for ROS?
Usually only one thread/process should communicate with PyBullet at one time, so you need to use some synchronization. There are two shared memory channels open by default, so you can use the second from a different thread as well.

**Comment by oscar-lima** [comment]
> Thanks for the update, that looks great from a first glance!

Thanks! :D

> 
> What do mean by parallelization? Is this a C plugin for PyBullet or a Python plugin for ROS?

pybullet plugins? Is something that I invented, pseudocode goes like this:

```
while application_alive:
    --- apply plugins in sequence (no threading yet) ---
    e.g. read odometry and publish to ROS (odometry  plugin)
    e.g. read user joint command from ROS and apply setJointMotorControlArray(pos, vel or torque) to robot (control plugin)
    e.g. use rayTestBatch to query laser hits on pybullet and publish ROS laser msg (laser plugin)
    -------------------------------------------------------------------
    step simulation # (self.pb.stepSimulation())
```

Note all pybullet plugins share the same import, for that I use importlib like this:

```
self.pb = importlib.import_module('pybullet')
```

and then I share self.pb to dynamically loaded plugins (also using importlib iterating over the required plugins)

> Usually only one thread/process should communicate with PyBullet at one time, so you need to use some synchronization. There are two shared memory channels open by default, so you can use the second from a different thread as well.

Here is where I have the main problem, because e.g. the laser task (rayTestBatch) takes a long to compute and slows down the flow. The synchronization you mention is something I would have to implement manually right? Can you provide with an example?
